src/configurationProvider.ts

```diff
@@ -0,0 +1,101 @@
+import * as fs from 'fs-extra';
+import * as path from 'path';
+import { parse } from 'jsonc-parser';
+import { OmniSharpServer } from './omnisharp/server';
+import * as serverUtils from './omnisharp/utils';
+import { AssetGenerator, addTasksJsonIfNecessary, createLaunchConfiguration, createAttachConfiguration, containsDotNetCoreProjects, createWebLaunchConfiguration } from './assets';
+import { isSubfolderOf } from './common';
+
+export class CSharpConfigurationProvider implements vscode.DebugConfigurationProvider {
+    private server: OmniSharpServer;
+
+    public constructor(server: OmniSharpServer) {
+        this.server = server;
+    }
+
+    /**
+     * TODO: Remove function when https://github.com/OmniSharp/omnisharp-roslyn/issues/909 is resolved.
+     * 
+     * Note: serverUtils.requestWorkspaceInformation only retrieves one folder for multi-root workspaces. Therefore, generator will be incorrect for all folders
+     * except the first in a workspace. Currently, this only works if the requested folder is the same as the server's solution path or folder.
+     */
+    private checkWorkspaceInformationMatchesWorkspaceFolder(folder: vscode.WorkspaceFolder | undefined): boolean {
+        const solutionPathOrFolder: string = this.server.getSolutionPathOrFolder();
+
+        // Make sure folder, folder.uri, and solutionPathOrFolder are defined.
+        if (!folder || !folder.uri || !solutionPathOrFolder)
+        {
+            return false;
+        }
+
+        let serverFolder = solutionPathOrFolder;
+        // If its a .sln file, get the folder of the solution.
+        if (fs.lstatSync(solutionPathOrFolder).isFile())
+        {
+            serverFolder = path.dirname(solutionPathOrFolder);
+        }
+
+        // Get absolute paths of current folder and server folder.
+        const currentFolder = path.resolve(folder.uri.fsPath);
+        serverFolder = path.resolve(serverFolder);
+
+        return currentFolder && folder.uri && isSubfolderOf(serverFolder, currentFolder);
+    }
+
+    /**
+	 * Returns a list of initial debug configurations based on contextual information, e.g. package.json or folder.
+	 */
+    provideDebugConfigurations(folder: vscode.WorkspaceFolder | undefined, token?: vscode.CancellationToken): vscode.ProviderResult<vscode.DebugConfiguration[]> {
+        return serverUtils.requestWorkspaceInformation(this.server).then(info => {
+            const generator = new AssetGenerator(info);
+            if (this.checkWorkspaceInformationMatchesWorkspaceFolder(folder) && containsDotNetCoreProjects(info)){
+                const dotVscodeFolder: string = path.join(folder.uri.fsPath, '.vscode');
+                const tasksJsonPath: string = path.join(dotVscodeFolder, 'tasks.json');
+                
+                // Make sure .vscode folder exists, addTasksJsonIfNecessary will fail to create tasks.json if the folder does not exist. 
+                fs.ensureDirSync(dotVscodeFolder);
```

Comment:
I'm seeing a lot of synchronous file I/O mixed with async code. Any concerns with that?

Reply:
I'm not too experienced with async programming patterns. I used the synchronous version of ensureDir and exist to make sure the .vscode folder exists and to make sure the .vscode/tasks.json does not exist before calling addTasksJsonIfNecessary. Is there a better way to do this?

Reply:
I don't think there's any concern for these small checks. I just wanted to make note of it.

Reply:
FWIW, NodeJS async isn't really challenging. Someday, maybe you can spend some time digging into it since the platform really hangs on asynchronous patterns. It's not necessary now.